File: Programmars/Level_2/뉴스 클러스터링.py
import math
import re
from collections import Counter


def split_string(string):
    strings = []
    string = string.upper()
    for i in range(len(string)-1):
        if string[i:i+2].isalpha():
            strings.append(string[i:i+2])
    return strings


def solution(str1, str2):
    count_str1 = Counter(split_string(str1))
    count_str2 = Counter(split_string(str2))

    union = 0
    intersection = 0
    for letters, cnt in count_str1.items():
        if letters in count_str2:
            intersection += min(cnt, count_str2[letters])
            union += max(cnt, count_str2[letters])
        else:
            union += cnt
    for letters, cnt in count_str2.items():
        if not letters in count_str1:
            union += cnt

    if intersection == union == 0:
        return 65536
    return int(intersection/union*65536)


print(solution('FRANCE', 'french'))  # 16384
print(solution('handshake', 'shake hands'))  # 65536
print(solution('aa1+aa2', 'AAAA12'))  # 43690
print(solution('E=M*C^2', 'e=m*c^2'))  # 65536


# ref: 정규표현식, set과 & 및 | 연산 활용
def solution(str1, str2):
    str1 = [str1[i:i+2].lower() for i in range(0, len(str1)-1)
            if not re.findall('[^a-zA-Z]+', str1[i:i+2])]
    str2 = [str2[i:i+2].lower() for i in range(0, len(str2)-1)
            if not re.findall('[^a-zA-Z]+', str2[i:i+2])]

    gyo = set(str1) & set(str2)
    hap = set(str1) | set(str2)

    if len(hap) == 0:
        return 65536

    gyo_sum = sum([min(str1.count(gg), str2.count(gg)) for gg in gyo])
    hap_sum = sum([max(str1.count(hh), str2.count(hh)) for hh in hap])

    return math.floor((gyo_sum/hap_sum)*65536)

File: JavaScript/Programmars/카카오블라인드/뉴스 클러스터링.js
const solution = (str1, str2) => {
  const isValidString = (str) => {
    return !/[^A-Z]/g.exec(str);
  };

  const splitString = (str) => {
    const strings = [];
    const len = str.length;
    for (let i = 0; i < len - 1; i++) {
      strings.push(str.toUpperCase().substr(i, 2));
    }

    return strings;
  };

  const makeCountMap = (set) => {
    const countMap = new Map();
    for (let str of set) {
      countMap.set(str, (countMap.get(str) || 0) + 1);
    }

    return countMap;
  };

  const countIntersection = (map1, map2) => {
    let count = 0;
    for (let [str1, count1] of map1) {
      count += Math.min(count1, map2.get(str1) || 0);
    }

    return count;
  };

  const countUnion = (map1, map2) => {
    const unionMap = map1;
    for (let [str2, count2] of map2) {
      if (unionMap.get(str2)) {
        unionMap.set(str2, Math.max(unionMap.get(str2), count2));
      } else {
        unionMap.set(str2, count2);
      }
    }
    let count = 0;
    for (let [, cnt] of unionMap) {
      count += cnt;
    }

    return count;
  };

  const set1 = splitString(str1).filter((str) => isValidString(str));
  const set2 = splitString(str2).filter((str) => isValidString(str));
  let answer = 65536;
  if (set1.length === 0 && set2.length === 0) return answer;

  const map1 = makeCountMap(set1);
  const map2 = makeCountMap(set2);
  const intersectionCount = countIntersection(map1, map2);
  const unionCount = countUnion(map1, map2);

  return Math.floor((intersectionCount / unionCount) * answer);
};

console.log(solution('FRANCE', 'french')); // 16384
console.log(solution('handshake', 'shake hands')); // 65536
console.log(solution('aa1+aa2', 'AAAA12')); // 43690
console.log(solution('E=M*C^2', 'e=m*c^2')); // 65536
